pop/service/service.go

```diff
@@ -97,124 +149,314 @@ func (s *Service) PinRequest(req *PinRequest) (network.Message, onet.ClientError
-	fs, ok := msg.(*FinalStatement)
+/* ------------InterConode Messages ----------- */
+
+// MergeConfig receives a final statement of requesting party,
+// hash of local party. Checks if they are from one merge party and responses with
+// own finalStatement
+func (s *Service) MergeConfig(req *network.Envelope) {
+	log.Lvlf2("%s gets MergeConfig from %s", s.Context.ServerIdentity().String(),
+		req.ServerIdentity.String())
+	mc, ok := req.Msg.(*mergeConfig)
 	if !ok {
-		log.Error("Couldn't convert to a FinalStatement")
+		log.Errorf("Didn't get a MergeConfig: %#v", req.Msg)
 		return
 	}
-	if err := fs.Verify(); err != nil {
-		log.Error(err)
+	if mc.Final == nil || mc.Final.Desc == nil {
+		log.Error("MergeConfig is empty")
 		return
 	}
-	s.data.Final = fs
-	s.save()
-	log.Lvlf3("%s Stored final statement %v", s.ServerIdentity(), s.data.Final)
+	mcr := &mergeConfigReply{PopStatusOK, mc.Final.Desc.Hash(), nil}
+
+	var final *FinalStatement
+	var m *merge
+	if final, ok = s.data.Finals[string(mc.ID)]; !ok {
+		log.Errorf("No config found")
+		mcr.PopStatus = PopStatusWrongHash
+		goto send
+	}
+	if m, ok = s.data.merges[string(mc.ID)]; !ok {
+		log.Error("No merge set found")
+		mcr.PopStatus = PopStatusWrongHash
+		goto send
+	}
+	if final.Verify() != nil {
+		log.Error("Local party's signature is invalid")
+		mcr.PopStatus = PopStatusMergeNonFinalized
+		goto send
+	}
+	mcr.PopStatus = final.VerifyMergeStatement(mc.Final)
+	if mcr.PopStatus < PopStatusOK {
+		goto send
+	}
+	if _, ok = m.statementsMap[string(mc.Final.Desc.Hash())]; ok {
+		log.Lvl2(s.ServerIdentity(), "Party was already merged, sent from",
+			req.ServerIdentity.String())
+		mcr.PopStatus = PopStatusMergeError
+		goto send
+	} else {
+		m.statementsMap[string(mc.Final.Desc.Hash())] = mc.Final
+	}
+
+	mcr.Final = final
+
+send:
+	err := s.SendRaw(req.ServerIdentity, mcr)
+	if err != nil {
+		log.Error("Couldn't send reply:", err)
+	}
+}
+
+// MergeConfigReply processes the response after MergeConfig message
+func (s Service) MergeConfigReply(req *network.Envelope) {
+	log.Lvlf2("MergeConfigReply: %s from %s got %v",
+		s.ServerIdentity(), req.ServerIdentity.String(), req.Msg)
+	mcrVal, ok := req.Msg.(*mergeConfigReply)
+	var mcr *mergeConfigReply
+	mcr = func() *mergeConfigReply {
+		if !ok {
+			log.Errorf("Didn't get a CheckConfigReply: %v", req.Msg)
+			return nil
+		}
+		var final *FinalStatement
+		if final, ok = s.data.Finals[string(mcrVal.PopHash)]; !ok {
+			log.Error("No party with given hash")
+			return nil
+		}
+		if mcrVal.PopStatus < PopStatusOK {
+			log.Error("Wrong pop-status:", mcrVal.PopStatus)
+			return mcrVal
+		}
+		if mcrVal.Final == nil {
+			log.Error("Empty FinalStatement in reply")
+			return nil
+		}
+		mcrVal.PopStatus = final.VerifyMergeStatement(mcrVal.Final)
+		return mcrVal
+	}()
+	if syncData, ok := s.syncs[string(mcrVal.PopHash)]; ok {
+		if len(syncData.mcChannel) == 0 {
+			syncData.mcChannel <- mcr
```

Comment:
As `mcr` is only used here, I'd put the above anonymous function inside the if, so it is not executed when it's not used.